### PR TITLE
Parse ALB X-Amzn-Trace-Id header into Honeycomb tracing fields

### PR DESCRIPTION
**Pull request by nathanleclaire**
Still not 100% sure this is a great idea, but figured it was worth posting my POC.

This will parse the `X-Amzn-Trace-Id` fields from the access log into Honeycomb tracing fields before sending the ALB events.

see for reference:

- https://docs.aws.amazon.com/xray/latest/devguide/xray-concepts.html
- https://docs.aws.amazon.com/elasticloadbalancing/latest/application/load-balancer-request-tracing.html

This might require a little bit of work on the app client side(s), to forward the `X-Amzn-Trace-Id` properly. I haven't tested it out fully yet, I've just been working off the documentation.

It won't really work with sampling, but deterministic sampling could be added too I suppose. This might be useful in the agentless integrations as well? What do you all think?

I erred on the side of keeping the old fields to retain backwards compat, etc. in this POC.

### CODE REVIEW

File: publisher/publisher_test.go [hunk 1]
```diff
@@ -8,6 +8,51 @@ import (
...
+				Data: map[string]interface{}{
+					"trace_id":                "Root=1-5759e988-bd862e3fe1be46a994272793;Parent=53995c3f42cd8ad8;Self=1-67891234-12456789abcdef012345678;Sampled=1",
+					"request_processing_time": 1.12,
+					"elb":          "fooservice",
+					"request_path": "/down/stream",
+				},
+			},
+			expected: map[string]interface{}{
+				// old fields
+				"trace_id":                "Root=1-5759e988-bd862e3fe1be46a994272793;Parent=53995c3f42cd8ad8;Self=1-67891234-12456789abcdef012345678;Sampled=1",
```
Comment (tredman) [inline publisher/publisher_test.go]: confused about these field names. We use `trace_id` as a special field in our trace UI. Won't this get picked up as the "root" span?

File: publisher/publisher_test.go [hunk 2]
```diff
@@ -8,6 +8,51 @@ import (
...
+				"request_processing_time": 1.12,
+				"elb":          "fooservice",
+				"request_path": "/down/stream",
+
+				"traceId":     "bd862e3fe1be46a994272793",
```
Comment (tredman) [inline publisher/publisher_test.go]: shouldn't these be `trace_id`, `span_id`, and `parent_id`?

File: publisher/publisher_test.go (hunk 1 above)
Comment (nathanleclaire) [reply publisher/publisher_test.go]: It's both `trace_id` and `traceId` I think :| I wonder if I should modify it when the trace fields are parsed

File: publisher/publisher_test.go (hunk 2 above)
Comment (nathanleclaire) [reply publisher/publisher_test.go]: I think both are valid but that would be more semantically consistent. I'll update

File: publisher/publisher.go [hunk 3]
```diff
@@ -149,6 +150,54 @@ func dropNegativeTimes(ev *event.Event) {
...
+		val := kv[1]
+		switch key {
+		case "Root":
+			versionTimeID := strings.Split(val, "-")
+			ev.Data["trace.trace_id"] = versionTimeID[2]
+		case "Self":
+			versionTimeID := strings.Split(val, "-")
+			ev.Data["trace.span_id"] = versionTimeID[2]
+			rootSpan = false
+		case "Parent":
```
Comment (maplebed) [inline publisher/publisher.go]: I don't see mention of `Parent` in the AWS docs. Can you say more here?

Comment (apechimp) [reply publisher/publisher.go]: It's used in an [example in the X-Ray docs](https://docs.aws.amazon.com/xray/latest/devguide/xray-concepts.html#xray-concepts-tracingheader).

Comment (nathanleclaire) [reply publisher/publisher.go]: Yea unfortunately I think making full use of this will be more complicated than just forwarding the headers directly.

**Pull request merged**